### ISSUE TITLE
Global bundle adjustment with CUDA slower than CPU

### ISSUE THREAD

**Issue by Xusj0w0** [colmap/colmap#2968]
**Describe the bug**
Low GPU utilization and GPU memory. Takes about 1h to register 1000 images, which is much slower than purely using CPU. The whole collection contains about 4300 images.
I have built ceres-solver and colmap with CUDA and CUDSS support.
May I ask what the "normal" GPU performance should be? How could I speed up colmap with GPU? Thank you.

**To Reproduce**
1. Extract features and use sequential_matcher with loop detection.
2. Run mapper with `--Mapper.ba_use_gpu=1`.

**Expected behavior**
Faster global BA than CPU.

**Screenshots**
Build info:
![image](https://github.com/user-attachments/assets/bc7c3d7b-9678-4813-95e6-28c9ac15eadd)
Sparse reconstruction log (global BA started at 19:04 and ended at 19:24):
![image](https://github.com/user-attachments/assets/eaa39a9b-96c6-40fc-ac53-11cba124f9fa)
GPU status (gpu memory increased from 19:06 and back to 270MB at 19:24):
![image](https://github.com/user-attachments/assets/42c00431-1df8-4f0a-9fb6-075e489cfe39)
![image](https://github.com/user-attachments/assets/31fd95c2-d619-4244-b263-5f109baa75a2)

**Environment:**
 - Ubuntu 20.04
 - COLMAP 3.11.0.dev0 -- Structure-from-Motion and Multi-View Stereo (Commit f5469f57 on 2024-11-26 with CUDA)
 - Capture Device: smartphone

**Comment by Xusj0w0** [comment]
> @Xusj0w0 Can share a Dockerfile that encapsulates your experiment?

Hello! I have uploaded the Dockerfile and the running logs.
- Regarding BA with CUDA support, I have uploaded the running logs on an Ubuntu 20.04 server. 
  - The dataset I used contains 4250 images of 3840x2160 resolution, and the reconstruction took 11 hours. 
  - It took about 30 hours purely using CPU in the same environment. The CUDA-based global BA is significantly faster than the CPU version when the image number exceeds 1500. 
  - I am not familiar with CUDSS and the sparse matrix solvers in Ceres. Is it normal for a reconstruction task of the above scale to use up to 3GB of GPU memory? Also, I would greatly appreciate it if you could provide some valuable suggestions for accelerating the reconstruction process.
- Also encountered an `SQLite error: "database disk image is malformed."` when running vocab_tree_matcher, which has occurred consistently before.
  - To reproduce, can run `image-malformed/sparse_rec.sh`

**Comment by mattiasmar** [comment]
@Xusj0w0  I haven't yet checked your attachment, but I did find these parameters which might be useful for you:
[Heuristic thresholds to switch from CPU to GPU usage](https://github.com/colmap/colmap/blob/e778be634029abbe14a3470dd4df6e53f475111f/src/colmap/estimators/bundle_adjustment.h#L140-L156)

**Comment by Xusj0w0** [comment]
> @Xusj0w0 I haven't yet checked your attachment, but I did find these parameters which might be useful for you: [Heuristic thresholds to switch from CPU to GPU usage](https://github.com/colmap/colmap/blob/e778be634029abbe14a3470dd4df6e53f475111f/src/colmap/estimators/bundle_adjustment.h#L140-L156)

Thank you for your reply. I will try increasing the number of images with the sparse GPU solver enabled.

**Comment by ahojnnes** [comment]
@mattiasmar @Xusj0w0 Yes, these thresholds are not tuned much (one of the reasons for this feature being disabled by default) and depend a lot on the specific GPU. You should be able to adjust the thresholds for your use case.

**Comment by ahojnnes** [comment]
@Xusj0w0 Regarding the issue with the database: Can it be that you mount the filesystem over the network?

**Comment by Xusj0w0** [comment]
> @Xusj0w0 Regarding the issue with the database: Can it be that you mount the filesystem over the network?

I use a local 16TB HDD. Can it be related to the SQLite version (3.31.0)?

**Comment by ahojnnes** [comment]
> > @Xusj0w0 Regarding the issue with the database: Can it be that you mount the filesystem over the network?
> 
> I use a local 16TB HDD. Can it be related to the SQLite version (3.31.0)?

No, this should not cause an issue.

**Comment by Xusj0w0** [comment]
@ahojnnes The above database error doesn't occur when I set the thread num to 1. Is it possible that there is a thread conflict?

**Comment by ahojnnes** [comment]
Can you show the full log, so we can see the context in which the error occurs? Thank you.

**Comment by ahojnnes** [comment]
OK, I see you already attached the logs. It can be a race condition but I thought I eradicated the last one. Are you sure you have these changes: https://github.com/colmap/colmap/commit/073f2e4e5049d62529cad871bea458ffe1a26b85 ?

**Comment by Xusj0w0** [comment]
@ahojnnes I built on commit [`f5469f57`](https://github.com/colmap/colmap/commit/f5469f57ae4a981f641c9b75836700585394dc7a). These changes have been applied.

**Comment by ahojnnes** [comment]
I am not sure what's going on. It doesn't look like a race condition when accessing the database. You could re-run the code after compiling with ASAN_ENABLED. Maybe it triggers something.

**Comment by ahojnnes** [comment]
This question has been addressed in #3163. GPU BA can be slower than CPU for small-to-medium reconstructions due to transfer overhead. The activation thresholds (`min_num_images_gpu_*`) are GPU-dependent and may need tuning. Closing as answered.
